### PR TITLE
Remove email addresses from UPN query in KeyVault Access Policy CmdLets.

### PR DESCRIPTION
**Pull request by RandalliLama**
## Description

Before this change, when a UPN argument is passed to one of the vault policy cmdlets the query for object id would include email addresses as well.  After this change UPN will mean UPN and there will be a seperate email argument if the user wants to look up the object id by email address.

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [x] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyPSHelp-documentation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [x] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [x] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by RandalliLama** [comment]
Please remove @pomortaz and @adarce as the default reviewers and replace with @randallilama , @schaabs , and @dragav .  @pomortaz has moved off of SDK and PowerShell work and Scott, Dragos and myself are now responsible for Key Vault SDK and PowerShell work.

**Comment by cormacpayne** [comment]
@azuresdkci add to whitelist

**Comment by cormacpayne** [comment]
@RandalliLama it looks like the default reviewers (who are just suggestions) are set by GitHub: ["Suggested reviewers are based on git blame data"](https://help.github.com/articles/requesting-a-pull-request-review/)

Feel free to request a review from those who you have mentioned previously.

File: src/ResourceManager/KeyVault/Commands.KeyVault.Test/Scripts/RunKeyVaultTests.ps1 [hunk 1]
```diff
@@ -211,6 +211,12 @@ function Run-AllControlPlaneTests
 
     # Set-AzureRmKeyVaultAccessPolicy & Remove-AzureRmKeyVaultAccessPolicy tests.
     Run-TestProtected { Run-VaultTest { Test_SetRemoveAccessPolicyByUPN } "Test_SetRemoveAccessPolicyByUPN" } "Test_SetRemoveAccessPolicyByUPN"
+
+    <#
+    This test is disabled because it requires a user with an email address that matches their UPN.
+    Run-TestProtected { Run-VaultTest { Test_SetRemoveAccessPolicyByEmail } "Test_SetRemoveAccessPolicyByEmail" } "Test_SetRemoveAccessPolicyByEmail"
```
Comment (cormacpayne) [inline src/ResourceManager/KeyVault/Commands.KeyVault.Test/Scripts/RunKeyVaultTests.ps1]: @RandalliLama is there any way to enable this test so we can (1) ensure that the functionality added does work as intended, (2) ensure that there are no regressions in the future?

Comment (RandalliLama) [inline src/ResourceManager/KeyVault/Commands.KeyVault.Test/Scripts/RunKeyVaultTests.ps1]: These tests are currently not run under the recording framework.  I am still investigating why that is, and what I can do about it.  As far as I can tell these tests are run manually by devs before checkin.  Since they are not recorded, if this test was enabled by default, the developer would need to ensure that they have a test environment that has a user with an email address, and further, that the email address matched their UPN.  That's not easy to set up if you don't already have one.  Setting up a user with an email address requires having a user licensed for Exchange, or doing manual REST calls to AAD graph.

If these tests were being run under the recording framework then only one person would need to have the necessary test setup and all other runs would be on recorded data.  Alternatively, we could set up a permanent test environment with users that have the necessary pre-reqs.  It's also possible to automate this kind of user creation in our current, manual, test infrastructure. 

1. The right long term answer is to get the tests running in the recording framework.  However, it is non-trivial to do so.  I need some time to investigate what it takes to get us completely back in to the recording framework.  I'd rather not invest significant effort in a temporary solution.  I plan on looking in to fixing this the right way in Calcium.
2. Without this fix it's possible for a user creating a new vault to inadvertently ACL it for someone other than the person running the CmdLet.  While we show the policy created by the CmdLet as part of the output, if the person running it isn't paying attention they may end up with a policy that grants access to the vault that they didn't intend.  This should be a very rare occurrence, but I would still like to get this fix out soon.  Blocking the release of this fix on test infrastructure investments could significantly delay it.

The risk of having this test disabled is small.  The code that does the lookup by email is factored out in to it's own method that is not shared by any other code path.

I may have missed something, and of course I am open to other suggestions.

Comment (RandalliLama) [inline src/ResourceManager/KeyVault/Commands.KeyVault.Test/Scripts/RunKeyVaultTests.ps1]: @cormacpayne ^

Comment (RandalliLama) [inline src/ResourceManager/KeyVault/Commands.KeyVault.Test/Scripts/RunKeyVaultTests.ps1]: By the way, I ran the test before disabling it.  I have verified that the feature works as intended.

File: src/ResourceManager/KeyVault/Commands.KeyVault/help/Set-AzureRmKeyVaultAccessPolicy.md [hunk 2]
```diff
@@ -28,6 +28,13 @@ Set-AzureRmKeyVaultAccessPolicy [-VaultName] <String> [[-ResourceGroupName] <Str
  [<CommonParameters>]
 ```
 
+### ByUserMail
+```
+Set-AzureRmKeyVaultAccessPolicy [-VaultName] <String> [[-ResourceGroupName] <String>]
+ -Mail <String> [-PermissionsToKeys <String[]>] [-PermissionsToSecrets <String[]>]
+ [-PermissionsToCertificates <String[]>] [-PermissionsToStorage <String[]>] [-PassThru] [-WhatIf] [-Confirm]
+ [<CommonParameters>]
+
```
Comment (adarce) [inline src/ResourceManager/KeyVault/Commands.KeyVault/help/Set-AzureRmKeyVaultAccessPolicy.md]: You need to close the triple backtick here:

    ```

**Comment by RandalliLama** [comment]
@cormacpayne I've addressed all your comments.  I've also enabled the new test by default.

**Comment by RandalliLama** [comment]
@adarce Thanks for the review.  I've addressed your comments as well.

File: src/ResourceManager/KeyVault/Commands.KeyVault/help/Set-AzureRmKeyVaultAccessPolicy.md [hunk 3]
```diff
@@ -28,6 +28,13 @@ Set-AzureRmKeyVaultAccessPolicy [-VaultName] <String> [[-ResourceGroupName] <Str
  [<CommonParameters>]
 ```
 
+### ByUserEmailAddress
+```
+Set-AzureRmKeyVaultAccessPolicy [-VaultName] <String> [[-ResourceGroupName] <String>]
+ -EmailAddress <String> [-PermissionsToKeys <String[]>] [-PermissionsToSecrets <String[]>]
+ [-PermissionsToCertificates <String[]>] [-PermissionsToStorage <String[]>] [-PassThru] [-WhatIf] [-Confirm]
+ [<CommonParameters>]
+
```
Comment (adarce) [inline src/ResourceManager/KeyVault/Commands.KeyVault/help/Set-AzureRmKeyVaultAccessPolicy.md]: This is still missing the closing triple backticks. View the rendered markdown here:

https://github.com/RandalliLama/azure-powershell/blob/c741b45a190943b364685f6edebc982dd0517d23/src/ResourceManager/KeyVault/Commands.KeyVault/help/Set-AzureRmKeyVaultAccessPolicy.md#byuseremailaddress

Comment (RandalliLama) [inline src/ResourceManager/KeyVault/Commands.KeyVault/help/Set-AzureRmKeyVaultAccessPolicy.md]: @adarce I could only find a single instance of missing back ticks.  Are there more?

Comment (adarce) [inline src/ResourceManager/KeyVault/Commands.KeyVault/help/Set-AzureRmKeyVaultAccessPolicy.md]: @RandalliLama That's it as far as what you introduced in to this PR. However, it looks like there's some more broken markdown in the same file that got introduced when PermissionsToStorage was added.

There needs to be a couple of newlines after "service principal." and before the triple backticks here:
https://github.com/RandalliLama/azure-powershell/blame/8a0fc06eadee571cc013612cd6d90f1ff0248998/src/ResourceManager/KeyVault/Commands.KeyVault/help/Set-AzureRmKeyVaultAccessPolicy.md#L370

**Markdown:** https://github.com/RandalliLama/azure-powershell/blob/8a0fc06eadee571cc013612cd6d90f1ff0248998/src/ResourceManager/KeyVault/Commands.KeyVault/help/Set-AzureRmKeyVaultAccessPolicy.md#-permissionstostorage

**Comment by cormacpayne** [comment]
@RandalliLama this is good to merge from our side; just need sign-off from your team.

Once this gets merged, let's open an issue to track enabling the KeyVault tests.

File: src/ResourceManager/KeyVault/AzureRM.KeyVault.psd1 [hunk 4]
```diff
@@ -12,7 +12,7 @@
 # RootModule = ''
 
 # Version number of this module.
-ModuleVersion = '3.1.0'
+ModuleVersion = '3.2.0'
```
Comment (cormacpayne) [inline src/ResourceManager/KeyVault/AzureRM.KeyVault.psd1]: @RandalliLama we will bump the versions during the release, so you can revert this change

File: src/ResourceManager/KeyVault/Commands.KeyVault.Test/Properties/AssemblyInfo.cs [hunk 5]
```diff
@@ -40,13 +40,13 @@
...
 //      Build Number
 //      Revision
 //
-// You can specify all the values or you can default the Build and Revision Numbers 
+// You can specify all the values or you can default the Build and Revision Numbers
 // by using the '*' as shown below:
 
 [assembly: AssemblyVersion( "3.0.0.0" )]
-[assembly: AssemblyFileVersion("3.1.0")]
+[assembly: AssemblyFileVersion("3.2.0")]
```
Comment (cormacpayne) [inline src/ResourceManager/KeyVault/Commands.KeyVault.Test/Properties/AssemblyInfo.cs]: @RandalliLama please revert this change to the version

File: src/ResourceManager/KeyVault/Commands.KeyVault/Properties/AssemblyInfo.cs [hunk 6]
```diff
@@ -29,8 +29,8 @@
 [assembly: CLSCompliant(false)]
 
 [assembly: Guid("2994548F-69B9-4DC2-8D19-52CC0C0C85BC")]
-[assembly: AssemblyVersion("3.1.0")]
-[assembly: AssemblyFileVersion("3.1.0")]
+[assembly: AssemblyVersion("3.2.0")]
+[assembly: AssemblyFileVersion("3.2.0")]
```
Comment (cormacpayne) [inline src/ResourceManager/KeyVault/Commands.KeyVault/Properties/AssemblyInfo.cs]: @RandalliLama please revert the changes made to version numbers in this file
